code/modules/integrated_electronics/subtypes/manipulation.dm

```diff
@@ -0,0 +1,421 @@
+	w_class = WEIGHT_CLASS_TINY
+	complexity = 10
+	inputs = list("target" = IC_PINTYPE_REF,"mode" = IC_PINTYPE_NUMBER)
+	outputs = list()
+	activators = list("pulse in" = IC_PINTYPE_PULSE_IN,"pulse out"=IC_PINTYPE_PULSE_OUT)
+	spawn_flags = IC_SPAWN_RESEARCH
+	power_draw_per_use = 50
+
+/obj/item/integrated_circuit/manipulation/plant_module/do_work()
+	..()
+	var/turf/T = get_turf(src)
+	var/obj/machinery/hydroponics/AM = get_pin_data_as_type(IC_INPUT, 1, /obj/machinery/hydroponics)
+	if(!istype(AM)) //Invalid input
+		return
+	var/mob/living/M = get_turf(AM)
+	if(!M.Adjacent(T))
+		return //Can't reach
+	switch(get_pin_data(IC_INPUT, 2))
+		if(0)
+			if(AM.myseed)
+				if(AM.harvest)
+					AM.myseed.harvest()
+					AM.harvest = 0
+					AM.lastproduce = AM.age
+					if(!AM.myseed.get_gene(/datum/plant_gene/trait/repeated_harvest))
+						qdel(AM.myseed)
+						AM.myseed = null
+						AM.dead = 0
+					AM.update_icon()
+		if(1)
+			AM.weedlevel = 0
+		if(2)
+			if(AM.myseed) //Could be that they're just using it as a de-weeder
+				AM.age = 0
+				AM.plant_health = 0
+				if(AM.harvest)
+					AM.harvest = FALSE //To make sure they can't just put in another seed and insta-harvest it
+				qdel(AM.myseed)
+				AM.myseed = null
+			AM.weedlevel = 0 //Has a side effect of cleaning up those nasty weeds
+			AM.update_icon()
+	activate_pin(2)
+
+/obj/item/integrated_circuit/manipulation/grabber
+	name = "grabber"
+	desc = "A circuit with it's own inventory for small/medium items, used to grab and store things."
+	icon_state = "grabber"
+	extended_desc = "The circuit accepts a reference to thing to be grabbed. It can store up to 10 things. Modes: 1 for grab. 0 for eject the first thing. -1 for eject all."
+	w_class = WEIGHT_CLASS_SMALL
+	size = 3
+
+	complexity = 10
+	inputs = list("target" = IC_PINTYPE_REF,"mode" = IC_PINTYPE_NUMBER)
+	outputs = list("first" = IC_PINTYPE_REF, "last" = IC_PINTYPE_REF, "amount" = IC_PINTYPE_NUMBER)
+	activators = list("pulse in" = IC_PINTYPE_PULSE_IN,"pulse out" = IC_PINTYPE_PULSE_OUT)
+	spawn_flags = IC_SPAWN_RESEARCH
+	power_draw_per_use = 50
+	var/max_w_class = WEIGHT_CLASS_NORMAL
+	var/max_items = 10
+/*
+/obj/item/integrated_circuit/manipulation/thrower/New()
+	processing_objects |= src
+
+/obj/item/integrated_circuit/manipulation/thrower/Destroy()
+	processing_objects -= src
+
+/obj/item/integrated_circuit/manipulation/thrower/process()
+	set_pin_data(IC_OUTPUT, 1, WEAKREF(contents[1]))
+	set_pin_data(IC_OUTPUT, 2, WEAKREF(contents[contents.len]))
+	set_pin_data(IC_OUTPUT, 3, contents.len)
+	push_data()
+*/
+/obj/item/integrated_circuit/manipulation/grabber/do_work()
+	var/turf/T = get_turf(src)
+	var/obj/item/AM = get_pin_data_as_type(IC_INPUT, 1, /obj/item)
+	var/turf/P = get_turf(AM)
+	var/mode = get_pin_data(IC_INPUT, 2)
+	if(mode == 1)
+		if(P.Adjacent(T))
+			if((contents.len < max_items) && AM && (AM.w_class <= max_w_class))
+				AM.forceMove(src)
```

Comment:
You should also add CanReach check somewhere around here.

Reply:
No. Lack of can reach allow to juggle items inside grabber.

Reply:
You need to define DirectAccess on the assembly.

Reply:
How?

Reply:
Only on assembly,or on components too?

Reply:
call `assembly.CanReach(...)`
with direct access looking something like this : (obviously pseudocode)
```
assembly/DirectAccess(x)
 if(..())
  return TRUE
 if(x in contents)
  return TRUE
 for(var/circuit in circuits)
  if(x in circuit.contents)
   return TRUE
```

Reply:
Or just `x in GetAllContents()`

Reply:
nope.bad idea.not gonna work.